File: backend/app/services/questionnaire_validation.py
"""Validação de realismo dos dados do questionário (Reisolari).

O questionário alimenta todo o motor de dimensionamento: um único número sem
sentido (uma área negativa, uma casa com 7 000 kWh/mês, uma tarifa de 10 €/kWh)
contamina em silêncio todos os cálculos seguintes. Este módulo é a fonte única
de verdade sobre o que é *impossível* e o que é apenas *fora do normal*:

* **Limites rígidos** — valores impossíveis no contexto. São recusados de
  imediato (o schema levanta erro, a API responde 422). A positividade campo a
  campo já é garantida pelas restrições ``Field`` do Pydantic; aqui juntam-se os
  tetos superiores e a verificação de sinal, para a mesma função também servir o
  frontend.
* **Avisos suaves** — valores *possíveis* mas muito fora do esperado para um
  telhado em Portugal. Nunca bloqueiam; o frontend pergunta "tens a certeza?",
  explica o porquê e sugere rever o dado.

Tudo aqui é **puro** (primitivos a entrar, primitivos a sair) — sem importar o
schema, sem BD, sem rede — logo é totalmente testável e sem ciclos de import. O
texto está propositadamente na voz Reisolari; o Easy Solar mantém um módulo
gémeo com palavras diferentes.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Optional

# ── Limites de engenharia (iguais nos dois projetos; só o texto muda) ──
CONSUMPTION_ANNUAL_HARD_MAX = 1_000_000.0      # kWh/ano — teto absoluto
CONSUMPTION_ANNUAL_SOFT_LOW = 600.0            # ≈ 50 kWh/mês
CONSUMPTION_ANNUAL_SOFT_HIGH_HOME = 15_000.0   # habitação
CONSUMPTION_ANNUAL_SOFT_HIGH_BIZ = 150_000.0   # empresa

AREA_HARD_MAX = 100_000.0                       # m² (10 ha) — telhado irreal
AREA_SOFT_HIGH_HOME = 500.0                     # m²
AREA_SOFT_HIGH_BIZ = 5_000.0                    # m²
AREA_MIN_PANEL_M2 = 2.0                          # abaixo disto não cabe 1 painel

PRICE_HARD_MAX = 10.0                            # €/kWh — impossível
PRICE_SOFT_HIGH = 0.60                           # €/kWh
PRICE_SOFT_LOW = 0.05                            # €/kWh

BATTERY_HARD_MAX = 200_000.0                     # €
BATTERY_SOFT_HIGH = 30_000.0                     # €

BUDGET_HARD_MAX = 10_000_000.0                   # €
BUDGET_SOFT_LOW = 1_000.0                        # €
BUDGET_SOFT_HIGH = 500_000.0                     # €


@dataclass(frozen=True)
class Warning:
    """Aviso de realismo, não bloqueante, mostrado como um "tens a certeza?"."""

    field: str
    message: str


def annual_equivalent(consumption_kwh: float, consumption_period: str) -> float:
    """Passa o consumo a anual (os valores mensais são ×12)."""
    return consumption_kwh * (12 if consumption_period == "mensal" else 1)


def _fmt(value: float) -> str:
    """Inteiro compacto para as mensagens (ex.: 84000 -> '84 000')."""
    return f"{value:,.0f}".replace(",", " ")


def hard_limit_error(
    *,
    consumption_kwh: float,
    consumption_period: str = "anual",
    available_area_m2: float = 0.0,
    electricity_price_eur_kwh: float = 0.20,
    wants_battery: bool = False,
    battery_cost_eur: float = 0.0,
    usage_type: str = "habitacao",
    budget_eur: Optional[float] = None,
) -> Optional[str]:
    """Devolve a 1.ª violação de limite rígido, ou ``None`` se tudo for válido.

    Cobre sinal/finitude *e* os tetos absolutos, para a mesma verificação
    proteger a API (via schema) e ser espelhada no frontend.
    """
    # --- Consumo ---
    if consumption_kwh is None or not math.isfinite(consumption_kwh) or consumption_kwh <= 0:
        return "O consumo é obrigatório e tem de ser maior que zero."
    annual = annual_equivalent(consumption_kwh, consumption_period)
    if annual > CONSUMPTION_ANNUAL_HARD_MAX:
        return "Consumo fora do alcance do simulador. Confirma os kWh na tua fatura."

    # --- Área ---
    if available_area_m2 is None or not math.isfinite(available_area_m2) or available_area_m2 < 0:
        return "A área do telhado não pode ser um valor negativo."
    if available_area_m2 > AREA_HARD_MAX:
        return (
            "Área de telhado irreal — demasiado grande. "
            "Desenha de novo só o teu telhado no mapa."
        )

    # --- Preço ---
    if (
        electricity_price_eur_kwh is None
        or not math.isfinite(electricity_price_eur_kwh)
        or electricity_price_eur_kwh < 0
    ):
        return "O preço da eletricidade não pode ser negativo."
    if electricity_price_eur_kwh > PRICE_HARD_MAX:
        return "Preço por kWh impossível. Corrige o valor (€/kWh)."

    # --- Bateria ---
    if wants_battery:
        if battery_cost_eur is None or not math.isfinite(battery_cost_eur) or battery_cost_eur < 0:
            return "O custo da bateria não pode ser negativo."
        if battery_cost_eur > BATTERY_HARD_MAX:
            return "Custo da bateria fora do razoável. Corrige o valor."

    # --- Orçamento ---
    if budget_eur is not None:
        if not math.isfinite(budget_eur) or budget_eur < 0:
            return "O orçamento não pode ser negativo."
        if budget_eur > BUDGET_HARD_MAX:
            return "Orçamento fora do razoável. Corrige o valor."

    return None


def soft_warnings(
    *,
    consumption_kwh: float,
    consumption_period: str = "anual",
    available_area_m2: float = 0.0,
    electricity_price_eur_kwh: float = 0.20,
    wants_battery: bool = False,
    battery_cost_eur: float = 0.0,
    usage_type: str = "habitacao",
    budget_eur: Optional[float] = None,
) -> list[Warning]:
    """Avisos "tens a certeza?" para valores possíveis mas fora do normal.

    Pressupõe que os dados já passaram em :func:`hard_limit_error` (ou seja, são
    finitos e estão dentro dos tetos absolutos).
    """
    out: list[Warning] = []

    # --- Consumo ---
    annual = annual_equivalent(consumption_kwh, consumption_period)
    soft_high = (
        CONSUMPTION_ANNUAL_SOFT_HIGH_BIZ
        if usage_type == "empresa"
        else CONSUMPTION_ANNUAL_SOFT_HIGH_HOME
    )
    if 0 < annual < CONSUMPTION_ANNUAL_SOFT_LOW:
        out.append(
            Warning(
                "consumption_kwh",
                f"Indicaste {_fmt(annual)} kWh/ano — é muito pouco. Tens a certeza? "
                "Confirma se escolheste o período certo (mensal ou anual).",
            )
        )
    elif annual > soft_high:
        ref = (
            "o de uma empresa comum"
            if usage_type == "empresa"
            else "o de uma casa normal (cerca de 2 500–5 000 kWh/ano)"
        )
        out.append(
            Warning(
                "consumption_kwh",
                f"Indicaste {_fmt(annual)} kWh/ano — bem acima d{ref}. Tens a certeza? "
                "Se calhar trocaste o valor mensal pelo anual; senão, dá uma vista de olhos na fatura.",
            )
        )

    # --- Área ---
    area_high = AREA_SOFT_HIGH_BIZ if usage_type == "empresa" else AREA_SOFT_HIGH_HOME
    if 0 < available_area_m2 < AREA_MIN_PANEL_M2:
        out.append(
            Warning(
                "available_area_m2",
                f"Com {available_area_m2:.1f} m² não cabe um painel inteiro. "
                "Tens a certeza do contorno? Volta a desenhar o telhado.",
            )
        )
    elif available_area_m2 > area_high:
        out.append(
            Warning(
                "available_area_m2",
                f"{_fmt(available_area_m2)} m² é um telhado enorme. Tens a certeza? "
                "Confirma que o contorno no mapa é só o teu telhado.",
            )
        )

    # --- Preço ---
    if 0 <= electricity_price_eur_kwh < PRICE_SOFT_LOW:
        out.append(
            Warning(
                "electricity_price_eur_kwh",
                f"{electricity_price_eur_kwh:.2f} €/kWh é muito baixo. "
                "Tens a certeza? Vê o preço na fatura.",
            )
        )
    elif electricity_price_eur_kwh > PRICE_SOFT_HIGH:
        out.append(
            Warning(
                "electricity_price_eur_kwh",
                f"{electricity_price_eur_kwh:.2f} €/kWh está acima do normal em Portugal "
                "(à volta de 0,15–0,25 €/kWh). Tens a certeza? Confere a fatura.",
            )
        )

    # --- Bateria ---
    if wants_battery and battery_cost_eur > BATTERY_SOFT_HIGH:
        out.append(
            Warning(
                "battery_cost_eur",
                f"{_fmt(battery_cost_eur)} € de bateria é muito dinheiro. "
                "Tens a certeza? Revê o orçamento da bateria.",
            )
        )

    # --- Orçamento ---
    if budget_eur is not None:
        if 0 < budget_eur < BUDGET_SOFT_LOW:
            out.append(
                Warning(
                    "budget_eur",
                    f"{_fmt(budget_eur)} € dá para pouco num sistema solar completo. "
                    "Podes continuar, mas as propostas podem ultrapassar este valor.",
                )
            )
        elif budget_eur > BUDGET_SOFT_HIGH:
            out.append(
                Warning(
                    "budget_eur",
                    f"{_fmt(budget_eur)} € é um orçamento muito acima do normal para uma "
                    "casa. Tens a certeza?",
                )
            )

    return out

File: frontend/lib/questionnaireValidation.ts
/**
 * Validação de realismo dos dados do questionário (Reisolari).
 *
 * Espelha, no cliente, os mesmos limites do backend
 * (`app/services/questionnaire_validation.py`) para dar resposta imediata:
 *
 *  - `hardErrors`  → valores impossíveis (negativos / absurdos). Bloqueiam o
 *    avanço e aparecem a vermelho junto ao campo. O backend aplica os mesmos
 *    tetos e devolve 422 a qualquer payload que escape ao cliente.
 *  - `softWarnings` → valores possíveis mas fora do normal para um telhado em
 *    Portugal. Não bloqueiam; perguntamos ao utilizador "Tens a certeza?", com a
 *    explicação do porquê e o convite a rever o dado.
 *
 * O texto está na voz Reisolari (tratamento informal), distinto do Easy Solar.
 */

import { Questionnaire } from "@/lib/questionnaire";

export const LIMITS = {
  CONSUMPTION_ANNUAL_HARD_MAX: 1_000_000,
  CONSUMPTION_ANNUAL_SOFT_LOW: 600,
  CONSUMPTION_ANNUAL_SOFT_HIGH_HOME: 15_000,
  CONSUMPTION_ANNUAL_SOFT_HIGH_BIZ: 150_000,
  AREA_HARD_MAX: 100_000,
  AREA_SOFT_HIGH_HOME: 500,
  AREA_SOFT_HIGH_BIZ: 5_000,
  AREA_MIN_PANEL_M2: 2,
  PRICE_HARD_MAX: 10,
  PRICE_SOFT_HIGH: 0.6,
  PRICE_SOFT_LOW: 0.05,
  BATTERY_HARD_MAX: 200_000,
  BATTERY_SOFT_HIGH: 30_000,
  BUDGET_HARD_MAX: 10_000_000,
  BUDGET_SOFT_LOW: 1_000,
  BUDGET_SOFT_HIGH: 500_000
} as const;

export type FieldKey =
  | "consumption_kwh"
  | "available_area_m2"
  | "electricity_price_eur_kwh"
  | "battery_cost_eur"
  | "budget_eur";

export type FieldErrors = Partial<Record<FieldKey, string>>;
export type Warning = { field: FieldKey; label: string; message: string };

/** Que passo (0-indexado) é dono de cada campo, para gating e navegação. */
export const FIELD_STEP: Record<FieldKey, number> = {
  consumption_kwh: 0,
  available_area_m2: 1,
  electricity_price_eur_kwh: 3,
  battery_cost_eur: 3,
  budget_eur: 3
};

const FIELD_LABEL: Record<FieldKey, string> = {
  consumption_kwh: "Consumo",
  available_area_m2: "Área do telhado",
  electricity_price_eur_kwh: "Preço da eletricidade",
  battery_cost_eur: "Custo da bateria",
  budget_eur: "Orçamento"
};

const isNum = (n: unknown): n is number => typeof n === "number" && Number.isFinite(n);

/** Inteiro com separador de milhares por espaço (ex.: 84000 -> "84 000"). */
function fmt(n: number): string {
  return Math.round(n)
    .toString()
    .replace(/\B(?=(\d{3})+(?!\d))/g, " ");
}

export function annualEquivalent(q: Questionnaire): number {
  return q.consumption_kwh * (q.consumption_period === "mensal" ? 12 : 1);
}

/** Erros bloqueantes (impossíveis): sinal, finitude e tetos absolutos. */
export function hardErrors(q: Questionnaire): FieldErrors {
  const e: FieldErrors = {};
  const annual = annualEquivalent(q);

  if (!isNum(q.consumption_kwh) || q.consumption_kwh <= 0) {
    e.consumption_kwh = "O consumo é obrigatório e tem de ser maior que zero.";
  } else if (annual > LIMITS.CONSUMPTION_ANNUAL_HARD_MAX) {
    e.consumption_kwh = "Consumo fora do alcance do simulador. Confirma os kWh na tua fatura.";
  }

  if (!isNum(q.available_area_m2) || q.available_area_m2 < 0) {
    e.available_area_m2 = "A área do telhado não pode ser um valor negativo.";
  } else if (q.available_area_m2 > LIMITS.AREA_HARD_MAX) {
    e.available_area_m2 = "Área de telhado irreal — demasiado grande. Desenha de novo só o teu telhado.";
  }

  if (!isNum(q.electricity_price_eur_kwh) || q.electricity_price_eur_kwh < 0) {
    e.electricity_price_eur_kwh = "O preço da eletricidade não pode ser negativo.";
  } else if (q.electricity_price_eur_kwh > LIMITS.PRICE_HARD_MAX) {
    e.electricity_price_eur_kwh = "Preço por kWh impossível. Corrige o valor (€/kWh).";
  }

  if (q.wants_battery) {
    if (!isNum(q.battery_cost_eur) || q.battery_cost_eur < 0) {
      e.battery_cost_eur = "O custo da bateria não pode ser negativo.";
    } else if (q.battery_cost_eur > LIMITS.BATTERY_HARD_MAX) {
      e.battery_cost_eur = "Custo da bateria fora do razoável. Corrige o valor.";
    }
  }

  if (q.budget_eur != null) {
    if (!isNum(q.budget_eur) || q.budget_eur < 0) {
      e.budget_eur = "O orçamento não pode ser negativo.";
    } else if (q.budget_eur > LIMITS.BUDGET_HARD_MAX) {
      e.budget_eur = "Orçamento fora do razoável. Corrige o valor.";
    }
  }

  return e;
}

/** Avisos "Tens a certeza?" para valores possíveis mas fora do normal. */
export function softWarnings(q: Questionnaire): Warning[] {
  const out: Warning[] = [];
  const push = (field: FieldKey, message: string) =>
    out.push({ field, label: FIELD_LABEL[field], message });

  const errs = hardErrors(q);
  const annual = annualEquivalent(q);

  // --- Consumo ---
  if (!errs.consumption_kwh) {
    const softHigh =
      q.usage_type === "empresa"
        ? LIMITS.CONSUMPTION_ANNUAL_SOFT_HIGH_BIZ
        : LIMITS.CONSUMPTION_ANNUAL_SOFT_HIGH_HOME;
    if (annual > 0 && annual < LIMITS.CONSUMPTION_ANNUAL_SOFT_LOW) {
      push(
        "consumption_kwh",
        `Indicaste ${fmt(annual)} kWh/ano — é muito pouco. Tens a certeza? Confirma se escolheste o período certo (mensal ou anual).`
      );
    } else if (annual > softHigh) {
      const ref =
        q.usage_type === "empresa"
          ? "o de uma empresa comum"
          : "o de uma casa normal (cerca de 2 500–5 000 kWh/ano)";
      push(
        "consumption_kwh",
        `Indicaste ${fmt(annual)} kWh/ano — bem acima d${ref}. Tens a certeza? Se calhar trocaste o valor mensal pelo anual; senão, dá uma vista de olhos na fatura.`
      );
    }
  }

  // --- Área ---
  if (!errs.available_area_m2) {
    const areaHigh =
      q.usage_type === "empresa" ? LIMITS.AREA_SOFT_HIGH_BIZ : LIMITS.AREA_SOFT_HIGH_HOME;
    if (q.available_area_m2 > 0 && q.available_area_m2 < LIMITS.AREA_MIN_PANEL_M2) {
      push(
        "available_area_m2",
        `Com ${q.available_area_m2.toFixed(1)} m² não cabe um painel inteiro. Tens a certeza do contorno? Volta a desenhar o telhado.`
      );
    } else if (q.available_area_m2 > areaHigh) {
      push(
        "available_area_m2",
        `${fmt(q.available_area_m2)} m² é um telhado enorme. Tens a certeza? Confirma que o contorno no mapa é só o teu telhado.`
      );
    }
  }

  // --- Preço ---
  if (!errs.electricity_price_eur_kwh) {
    if (q.electricity_price_eur_kwh >= 0 && q.electricity_price_eur_kwh < LIMITS.PRICE_SOFT_LOW) {
      push(
        "electricity_price_eur_kwh",
        `${q.electricity_price_eur_kwh.toFixed(2)} €/kWh é muito baixo. Tens a certeza? Vê o preço na fatura.`
      );
    } else if (q.electricity_price_eur_kwh > LIMITS.PRICE_SOFT_HIGH) {
      push(
        "electricity_price_eur_kwh",
        `${q.electricity_price_eur_kwh.toFixed(2)} €/kWh está acima do normal em Portugal (à volta de 0,15–0,25 €/kWh). Tens a certeza? Confere a fatura.`
      );
    }
  }

  // --- Bateria ---
  if (!errs.battery_cost_eur && q.wants_battery && q.battery_cost_eur > LIMITS.BATTERY_SOFT_HIGH) {
    push(
      "battery_cost_eur",
      `${fmt(q.battery_cost_eur)} € de bateria é muito dinheiro. Tens a certeza? Revê o orçamento da bateria.`
    );
  }

  // --- Orçamento ---
  if (!errs.budget_eur && q.budget_eur != null) {
    if (q.budget_eur > 0 && q.budget_eur < LIMITS.BUDGET_SOFT_LOW) {
      push(
        "budget_eur",
        `${fmt(q.budget_eur)} € dá para pouco num sistema solar completo. Podes continuar, mas as propostas podem ultrapassar este valor.`
      );
    } else if (q.budget_eur > LIMITS.BUDGET_SOFT_HIGH) {
      push(
        "budget_eur",
        `${fmt(q.budget_eur)} € é um orçamento muito acima do normal para uma casa. Tens a certeza?`
      );
    }
  }

  return out;
}

/** Erros bloqueantes só dos campos de um dado passo (gating do "Seguinte"). */
export function stepErrors(q: Questionnaire, step: number): FieldErrors {
  const all = hardErrors(q);
  const out: FieldErrors = {};
  (Object.keys(all) as FieldKey[]).forEach((f) => {
    if (FIELD_STEP[f] === step) out[f] = all[f];
  });
  return out;
}
